### ISSUE TITLE
Adding ALLEGREX architecture to snowman

### ISSUE THREAD

**Issue by hlide** [yegord/snowman#36]
I may plan to add a new architecture instead of using the current MIPS architecture which is a work in progress. because ALLEGREX is not recognized by capstone framework and the fact the latter is using LLVM make the implementation of ALLEGREX too complex. There are subtle differences which make the use of MIPS architecture not viable for decompiling ALLEGREX code.

So, I am about to provide a specific disassembler (the one provided by pspdecompiler, based on prxtools one, but with addition of a decomposer) for the architecture analyser. It means handling of all instructions including VFPU.

I am pretty sure people from uOFW project may be interested as well. But I am also pretty sure that it will be hard to decompile kernel modules because they may use some tricks which are not ABI compliant, so I am expecting more task to do.

As for the author in http://lists.derevenets.com/pipermail/snowman/2015-August/000002.html, it may be great that he/she contributes as well here (PRX handling).

**Comment by hlide** [comment]
By the way, if a MIPS code is accessing a memory slot which is not a variable in the stack, it should be imperatively kept in the c++. Especially, kernel modules have a large range of address for MMIO, the controllers of which may be triggered even if the access may look like a dummy one by the decompiler. Decompiler should not be smarter than  the programmer because it does not know the context of such memory access.

**Comment by yegord** [comment]
Following you reasoning, one can go far. :-) For example, the decompiler should not reconstruct expressions then: reconstruction of expressions involves reordering of operations, including memory accesses, which may change the semantics of the program. We can then just print IR in a C form (like LLVM C backend did, before it was removed), and go home. :-)

My proposal is to first find such examples, and then discuss what to do with them.

P.S. Assume you have a load to `r1` from address `r2` that is not used. What do you want to be generated for it? `r1 = *r2;`? Just `*r2;`?

**Comment by hlide** [comment]
`*r2;` is fine but I would expect something more like `*((int8_t*)0xb8c0004);` which could mean sending an ack to controller by reading at this address OR just reading a byte in a FIFO queue and discard it because it is just a byte we don't care. What I mean is if a memory access as a read or a write is done and is not relative to a stack, there is a reason to be here and to keep it. As for register or stack slots being assigned with a value of a memory access, they can be discarded but not the memory access. I may be wrong with my impression, but I had the impression snowman is discarding the memory access as well if it is not propagated, as if it was the same as a register.

**Comment by hlide** [comment]
And yes, you SHOULD never reorder the memory accesses (i mean those not involving with variable in stack) - remember a case I gave you where there was like an exchange between two memory slots : c++ result was wrong because it reverses order, preventing to keep the old value in safe place.

**Comment by yegord** [comment]
One can here https://github.com/yegord/snowman/blob/master/src/nc/core/ir/cgen/DefinitionGenerator.cpp#L636, instead of returning null, recursively traverse `assignment->right()`, find there all memory accesses that you want to preserve, aggregate them into a single expression using, say, operator comma, and return an ExpressionStatement with this expression. If you really want this feature so much, you can prepare a patch.

**Comment by hlide** [comment]
Do memory accesses to variable in stack are transformed into some virtual registers so I can distinguish them from the other memory accesses and keep only the latter?

**Comment by hlide** [comment]
Is there any reason to return a big expression with operator comma instead of  a block of expressions? I mean the left value is discarded (`lhs = rhs`-> `rhs`). Oh you mean something like:
`lhs = f(rhs1, rhs2, rhs3);` --> `rhs1, rhs2;` instead of `f(rhs1, rhs2, rhs3);` when *rhs1* and *rhs2* are memory access to keep and lhs is discardable?

**Comment by yegord** [comment]
On Wed, Aug 05, 2015 at 02:59:02PM -0700, hlide wrote:
> Do memory accesses to variable in stack are transformed into some
> virtual registers so I can distinguish them from the other memory
> accesses and keep only the latter?

No. But you can query the memory location of a given term by calling
dataflow_.getMemoryLocation(term). By looking at the location's domain
you can find out, where the object identified by the term is: on the
stack, in the global memory, in a register (see MemoryDomain.h).

-- 
Yegor Derevenets

**Comment by hlide** [comment]
Ok, I will not work on it at once. Priority is to make the Allegrex analyzer to work and implement every instructions as many as possible.

**Comment by yegord** [comment]
On Wed, Aug 05, 2015 at 03:06:28PM -0700, hlide wrote:
> Oh you mean something like: `lhs = f(rhs1, rhs2, rhs3);` --> `rhs1,
> rhs2;` instead of `f(rhs1, rhs2, rhs3);` when *rhs1* and *rhs2* are
> memory access to keep and lhs is discardable?
Yes.

In practice, the commas should be rarely needed.

You can keep the whole expression, if this is what you want to.

Ah, and yes, you will have to modify LivenessAnalyzer to call makeLive
on the terms representing these rhs1, rhs2, rhs3. But it is no harder
than modifying DefinitionGenerator.

And one more thing: for recursive descent through the terms you can use
Term::callOnChildren().

-- 
Yegor Derevenets

**Comment by nihilus** [comment]
Not reading thru all the post I'd say that @hlide is on to something and I'm to kinda subset of this 'something'

Bringing in the ALLEGREX ISA is more than welcomed by me even if it necessary means split up mean and @hlide into two diversions (I prefer this since it is not like soccers with it's divisions etc :-)). I rest assure that @hlide will help and improve my (like he/she has done all the way down )understanding and vice-versa.

**Comment by nihilus** [comment]
@yegord @hlide ... hate to admit this and it is 'off the record' but of us all @hlide choosed the most love avatar,... 

@hlide  you indicated before that you are in the central European timezone, is this correct (one of my brother's still use free.fr when working in Provence)?

**Comment by hlide** [comment]
@yegord I'm in France and a french man despite of my avatar :-).

As for ALLEGREX, there are too many *dialects* for MIPS and *capstone* is pretty irritating and lacking when not used as a simple disassembler. Beside, VFPU is too specific and complex to handle it as a *dialect* for MIPS (VFPU is an awesome kind of a proprietary coprocessor). But MIPS and ALLEGREX will benefit together.

**Comment by hlide** [comment]
Ok, disassembler seems ok:

![image](https://cloud.githubusercontent.com/assets/56785/9110178/d7d547fc-3c3b-11e5-8598-614498dda117.png)

*CPU*, *FPU* and *VFPU* instructions can be seen in this image. I also change the detection of architecture to be sure to select allegrex only if `elfhdr.e_flags`contains a special value to tell the machine is ALLEGREX, so MIPS architecture could still be used as a fallback.

**Comment by nihilus** [comment]
@hlide is that capstone?

**Comment by hlide** [comment]
nope, AllegrexDisassembler has its own disassembler.

**Comment by nihilus** [comment]
@hlide neat that you integrated it.

**Comment by hlide** [comment]
@yegord I have an issue with branch likely instructions (B*xx*L) when generating AST.

See this example:

```c++
                    case I_BEQL: {
                        AllegrexExpressionFactoryCallback then(factory, program->createBasicBlock(), instruction);
                        _[
                            jump(gpr(0) == gpr(1),
                                 (delayslot(then)[jump(imm(2))]).basicBlock(),
                                 directSuccessorButOne())
                        ];
                        break;
                    }
```

As you can see, this instruction NEVER jumps *directSuccessor* (*address + 4*) but to *directSuccessorButOne* (*address + 8*). The instruction at *address + 4* is inserted just before the jump as a branch delay slot when taken. I think it may fool the function *IRGenerator::addJumpToDirectSuccessor*.

**Comment by hlide** [comment]
@yegord more info here:

Source:
```c++
                    case I_BLTZL: {
                        AllegrexExpressionFactoryCallback then(factory, program->createBasicBlock(), instruction);
                        _[
                            jump(signed_(gpr(0)) < constant(0),
                                 (delayslot(then)[jump(imm(1))]).basicBlock(),
                                 directSuccessorButOne())
                        ];
                    }
```

![image](https://cloud.githubusercontent.com/assets/56785/9135698/7bfa212a-3d10-11e5-9d00-b7caf715bc94.png)

![image](https://cloud.githubusercontent.com/assets/56785/9135731/d59a16c2-3d10-11e5-8095-10c4ec651f51.png)

```diff
diff --git a/src/nc/core/ir/cgen/DefinitionGenerator.cpp b/src/nc/core/ir/cgen/DefinitionGenerator.cpp
index c81c88c..9567cd8 100644
--- a/src/nc/core/ir/cgen/DefinitionGenerator.cpp
+++ b/src/nc/core/ir/cgen/DefinitionGenerator.cpp
@@ -24,6 +24,7 @@
 
 #include "DefinitionGenerator.h"
 
+#include <nc/common/LogToken.h>
 #include <nc/common/Foreach.h>
 #include <nc/common/Range.h>
 #include <nc/common/Unreachable.h>
@@ -507,6 +508,12 @@ std::unique_ptr<likec::Expression> DefinitionGenerator::makeExpression(const cfl
             std::unique_ptr<likec::Expression> expression;
 
             if (const Jump *jump = statement->asJump()) {
+
+                if (jump != basicNode->basicBlock()->getJump())
+                {
+                    nc::LogToken::instance()->error(QString("%1: %2").arg(jump->instruction()->addr(), -8, 16).arg(jump->instruction()->toString()));
+                }
+
                 assert(jump == basicNode->basicBlock()->getJump());
 
                 expression = makeExpression(jump->condition());
@@ -710,7 +717,6 @@ std::unique_ptr<likec::Statement> DefinitionGenerator::doMakeStatement(const Sta
                                         parent().makeType(parent().types().getType(returnValueTerm)),
                                         std::move(callOperator))));
                         }
-                        }
                     }
                 }
             }
```

**Comment by hlide** [comment]
As expected, if I comment all branch likely instruction (B*xx*L), I can decompile:

![image](https://cloud.githubusercontent.com/assets/56785/9137705/c8be0b8e-3d1f-11e5-8370-2f1717629cac.png)

**Comment by nihilus** [comment]
@hlide that should go for MIPS as well then... :-/ However cool and fast work.

**Comment by hlide** [comment]
@nihilus yeah the issue is the same for MIPS.

**Comment by yegord** [comment]
On Fri, Aug 07, 2015 at 05:20:14AM -0700, hlide wrote:
> The instruction at *address + 4* is inserted just before the jump as a
> branch delay slot when taken. I think it may fool the function
> *IRGenerator::addJumpToDirectSuccessor*. 

ir::Program::createBasicBlock(ByteAddr) and cgen::DefinitionGenerator::
isDominating expect that the addresses of the instructions of the
statements in a single basic block constitute a non-descending sequence. 

When generating statements for the delay-slot instruction, you should
pass the preceding instruction to the factory callback, to make these
statements appear to originate from this preceding instruction.

-- 
Yegor Derevenets

**Comment by hlide** [comment]
is it now what I am already doing with:
```c++
AllegrexExpressionFactoryCallback then(factory, program->createBasicBlock(), instruction);
```
?

*instruction* is the preceding instruction of delayslot, that is, the branch instruction.

Just for you information, branch instruction *not unlikely* appear to work fine:

```c++
case I_BEQ: {
    AllegrexExpressionFactoryCallback then(factory, program->createBasicBlock(), instruction);
    _[
        jump(gpr(0) == gpr(1),
             (delayslot(then)[jump(imm(2))]).basicBlock(),
             directSuccessor())
    ];
    break;
}
```
while this one doesn't work:
```c++
case I_BEQL: {
    AllegrexExpressionFactoryCallback then(factory, program->createBasicBlock(), instruction);
    _[
        jump(gpr(0) == gpr(1),
             (delayslot(then)[jump(imm(2))]).basicBlock(),
             directSuccessorButOne())
    ];
    break;
}
```

the only diffrence is about *directSuccessor()* and *directSuccessorButOne()* as the last argument for *jump*.

**Comment by yegord** [comment]
On Fri, Aug 07, 2015 at 03:54:29PM -0700, hlide wrote:
> is it now what I am already doing with:
> ```c++
> AllegrexExpressionFactoryCallback then(factory, program->createBasicBlock(), instruction);
> ```
> ?
Yes. After rereading the code I got it.

I have pushed small well-formedness checks in 82f0a716. Do they fail?
If yes, it should be easy to find the problem source (e.g., print the
basic block with the problem). If no, you can give me the commit id, the
input file, and the steps to reproduce the problem — I can have a look.

-- 
Yegor Derevenets

**Comment by hlide** [comment]
Ok, I merged your changes. Now it also fails on *branch not likely* too.

I have this assert:
```c++
#ifndef NDEBUG
    /*
     * Check that jump is always the last instruction in a basic block.
     */
    foreach (auto basicBlock, program_->basicBlocks()) {
        foreach (auto statement, basicBlock->statements()) {
            if (auto jump = statement->asJump()) {
                if (jump != basicBlock->statements().back()) {
                    nc::LogToken::instance()->error(QString("%1: %2").arg(jump->instruction()->addr(), -8, 16).arg(jump->instruction()->toString()));
                    nc::LogToken::instance()->error(QString("jump: %1").arg(jump->toString()));
                    nc::LogToken::instance()->error(QString("basicBlock->statements().back(): %1").arg(basicBlock->statements().back()->toString()));
                }
                assert(jump == basicBlock->statements().back());
            }
        }
    }
#endif
```
Log:
```
[Info] Decompiling.
[Info] Creating intermediate representation of the program.
[Error] 890179c : blez       $s0, 0x08901844
[Error] jump: if (<1000:512..543> (signed)< 0x11) goto basic block 0x9402543ac0 else goto basic block 0x9402544a80

[Error] basicBlock->statements().back(): goto address 0x8901844

```
I dumped `jump` and `basicBlock->statements().back()`.

**Comment by yegord** [comment]
On Sat, Aug 08, 2015 at 01:12:08PM -0700, hlide wrote:
> Log:
> ```
> [Info] Decompiling.
> [Info] Creating intermediate representation of the program.
> [Error] 890179c : blez       $s0, 0x08901844
> [Error] jump: if (<1000:512..543> (signed)< 0x11) goto basic block 0x9402543ac0 else goto basic block 0x9402544a80
> 
> [Error] basicBlock->statements().back(): goto address 0x8901844
> 
> ```
> I dumped `jump` and `basicBlock->statements().back()`.

Can you print the whole basic block? I would like to see: all
statements, all instructions, whether the basic block has an address.

-- 
Yegor Derevenets

**Comment by hlide** [comment]
done after edited

**Comment by hlide** [comment]
always check my posts on github, I often edited them again (mistakes, more info, etc.)

**Comment by yegord** [comment]
Does the last instruction (the second jump) has the instruction set?

**Comment by hlide** [comment]
I do not know where it comes from. It is certainly not I who added it. I think it is the function *IRGenerator::addJumpToDirectSuccessor* that did it.

**Comment by yegord** [comment]
But it is not called by that time yet. :)
Let's see what is the instruction.

**Comment by hlide** [comment]
how can I have it? I am not fluent enough with all that stuff.

**Comment by yegord** [comment]
`basicBlock->statements()->back()->instruction()->toString()`

**Comment by hlide** [comment]
ah ok, `basicBlock->statements().back()->instruction()->toString()`, right ?

**Comment by yegord** [comment]
Yep. I got one arrow instead of dot wrong. :)

**Comment by hlide** [comment]
same instruction:
```
[Info] Decompiling.
[Info] Creating intermediate representation of the program.
[Error] 890179c : blez       $s0, 0x08901844
[Error] 890179c : blez       $s0, 0x08901844
[Error] jump: if (<1000:512..543> (signed)< 0x11) goto basic block 0xb48b3fed90 else goto basic block 0xb48b3fe760

[Error] basicBlock->statements().back(): goto address 0x8901844

[Error] basicBlock: basicBlock0xb48b3fed00 [shape=box,label="Address: None\n<1000:0..31> = 0x0\nif (<1000:512..543> (signed)< 0x11) goto basic block 0xb48b3fed90 else goto basic block 0xb48b3fe760\ngoto address 0x8901844\n"];
```

**Comment by yegord** [comment]
Then, you add it yourself, when handling this instruction. I'll have a look at your code.

**Comment by hlide** [comment]
https://github.com/hlide/snowman/blob/master/src/nc/arch/allegrex/AllegrexInstructionAnalyzer.cpp

**Comment by yegord** [comment]
Are you sure you do not have two jump instructions next to each other in the disassembly?

**Comment by hlide** [comment]
![image](https://cloud.githubusercontent.com/assets/56785/9152288/b6d75cbe-3e22-11e5-9a6c-cc905fb58c13.png)

**Comment by yegord** [comment]
Let's make the program cache at the moment of adding the second jump.

```diff
diff --git a/src/nc/core/irgen/Expressions.h b/src/nc/core/irgen/Expressions.h
index 38e7372..95e570e 100644
--- a/src/nc/core/irgen/Expressions.h
+++ b/src/nc/core/irgen/Expressions.h
@@ -1274,6 +1274,8 @@ public:
 
     void operator()(std::unique_ptr<ir::Statement> statement) const {
         statement->setInstruction(mInstruction);
+        assert(basicBlock()->address() != boost::none || basicBlock()->statements().empty() ||
+               !basicBlock()->statements().back()->isTerminator());
         mBasicBlock->pushBack(std::move(statement));
     }

```

**Comment by hlide** [comment]
here is what I get with this code addition:

```
[Info] Decompiling.
[Info] Creating intermediate representation of the program.
[Error] basicBlock0xd6d4959fd0 [shape=box,label="Address: None\n<1000:0..31> = 0x0\nif (<1000:512..543> (signed)< 0x11) goto basic block 0xd6d495ac30 else goto basic block 0xd6d495a8d0\n"];
```

**Comment by hlide** [comment]
`address: None` is the issue, right?

**Comment by yegord** [comment]
But what is the call stack, when this assertion fails?

Address: None is not an issue — it essentially means, that you created this basic block not using getBasicBlockForInstruction/createBasicBlock(ByteAddr), but by creating a fresh basic block using createBasicBlock().

**Comment by hlide** [comment]
Updated log above

![image](https://cloud.githubusercontent.com/assets/56785/9152412/ba7e104c-3e27-11e5-8c51-5a89a5156339.png)

**Comment by yegord** [comment]
Line numbers, the most interesting thing, are not visible.

**Comment by hlide** [comment]
```
 	msvcr120d.dll!_NMSG_WRITE(int rterrnum) Line 226	C
 	msvcr120d.dll!abort() Line 72	C
 	msvcr120d.dll!_wassert(const wchar_t * expr, const wchar_t * filename, unsigned int lineno) Line 380	C
 	snowman.exe!nc::core::irgen::expressions::ExpressionFactoryCallback<nc::arch::allegrex::`anonymous namespace'::AllegrexExpressionFactory>::operator()(std::unique_ptr<nc::core::ir::Statement,std::default_delete<nc::core::ir::Statement> > statement) Line 1285	C++
>	snowman.exe!nc::core::irgen::expressions::ExpressionFactoryCallback<nc::arch::allegrex::`anonymous namespace'::AllegrexExpressionFactory>::doCallback<nc::core::irgen::expressions::JumpStatement<nc::core::irgen::expressions::NullExpression,nc::core::irgen::expressions::TermExpression> >(nc::core::irgen::expressions::StatementBase<nc::core::irgen::expressions::JumpStatement<nc::core::irgen::expressions::NullExpression,nc::core::irgen::expressions::TermExpression> > & statement) Line 1296	C++
 	snowman.exe!nc::core::irgen::expressions::ExpressionFactoryCallback<nc::arch::allegrex::`anonymous namespace'::AllegrexExpressionFactory>::operator[]<nc::core::irgen::expressions::JumpStatement<nc::core::irgen::expressions::NullExpression,nc::core::irgen::expressions::TermExpression> >(nc::core::irgen::expressions::StatementBase<nc::core::irgen::expressions::JumpStatement<nc::core::irgen::expressions::NullExpression,nc::core::irgen::expressions::TermExpression> > && statement) Line 1274	C++
 	snowman.exe!nc::arch::allegrex::AllegrexInstructionAnalyzerImpl::createStatements(nc::core::irgen::expressions::ExpressionFactoryCallback<nc::arch::allegrex::`anonymous namespace'::AllegrexExpressionFactory> & _, const nc::arch::allegrex::AllegrexInstruction * instruction, nc::core::ir::Program * program) Line 379	C++
 	snowman.exe!nc::arch::allegrex::AllegrexInstructionAnalyzerImpl::createStatements(const nc::arch::allegrex::AllegrexInstruction * instruction, nc::core::ir::Program * program) Line 81	C++
 	snowman.exe!nc::arch::allegrex::AllegrexInstructionAnalyzer::doCreateStatements(const nc::core::arch::Instruction * instruction, nc::core::ir::Program * program) Line 1105	C++
 	snowman.exe!nc::core::irgen::InstructionAnalyzer::createStatements(const nc::core::arch::Instruction * instruction, nc::core::ir::Program * program) Line 42	C++
 	snowman.exe!nc::core::irgen::IRGenerator::generate() Line 78	C++
 	snowman.exe!nc::core::MasterAnalyzer::createProgram(nc::core::Context & context) Line 70	C++
 	snowman.exe!nc::core::MasterAnalyzer::decompile(nc::core::Context & context) Line 229	C++
 	snowman.exe!nc::core::Driver::decompile(nc::core::Context & context) Line 101	C++
 	snowman.exe!nc::gui::Decompilation::work() Line 45	C++
 	snowman.exe!nc::gui::Activity::run() Line 32	C++
 	Qt5Cored.dll!QThreadPoolThread::run() Line 93	C++
 	Qt5Cored.dll!QThreadPrivate::start(void * arg) Line 346	C++
 	msvcr120d.dll!_callthreadstartex() Line 376	C
 	msvcr120d.dll!_threadstartex(void * ptd) Line 359	C
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown
```

**Comment by yegord** [comment]
Can you split filling in `then` and `_` into two statements? This will bring more clarity.

My current hypothesis, is that `delayslot(then)` inserts the first jump, `[jump(imm(1))]` inserts the second. Maybe you could print the delayslot instruction?

**Comment by hlide** [comment]
I tried several approaches but with no avail:

```c++
    case I_BLEZ: {
        auto then = delayslot(AllegrexExpressionFactoryCallback(factory, program->createBasicBlock(), instruction));
        then[jump(imm(1))];
        _[
            jump(signed_(gpr(0)) <= constant(0),
                 then.basicBlock(),
                 directSuccessor())
        ];
        break;
```
and even without delayslot():

```
    case I_BLEZ: {
        auto then = AllegrexExpressionFactoryCallback(factory, program->createBasicBlock(), instruction);
        then[jump(imm(1))];
        _[
            jump(signed_(gpr(0)) <= constant(0),
                 then.basicBlock(),
                 directSuccessor())
        ];
        break;
```

**Comment by hlide** [comment]
same issue

**Comment by hlide** [comment]
How and when do I print the delayslot?

**Comment by yegord** [comment]
In the last code snippet, everything crashes inside `then[jump(imm(1))];`? Is `then.basicBlock()->statements().empty()` at all?

**Comment by hlide** [comment]
wait, I am checking again those tests - I did not check if it was on the same instruction.

**Comment by yegord** [comment]
delayslot instruction you can print using qDebug() or similar.

**Comment by hlide** [comment]
in `case I_BLEZ:`?

**Comment by yegord** [comment]
In delayslot lambda.

**Comment by yegord** [comment]
qDebug() << "And the delayslot instruction is..." << delayslot->toString();

**Comment by hlide** [comment]
without *delayslot*, *BLEZ* is fine (but obviously wrong in emulation) - no assert.

with *delayslot* even outside of `_[...]` still triggers the assert.

**Comment by yegord** [comment]
And the delayslot instruction is?..

**Comment by hlide** [comment]
![image](https://cloud.githubusercontent.com/assets/56785/9152531/2a139f76-3e2d-11e5-82d8-9d758b0ef8fd.png)

**Comment by yegord** [comment]
And InstructionAnalyzer for SLTI generates guess what? A jump!

**Comment by yegord** [comment]
Sorry for not spotting this an hour ago. :)

**Comment by hlide** [comment]
and why a jump?

**Comment by yegord** [comment]
https://github.com/hlide/snowman/blob/master/src/nc/arch/allegrex/AllegrexInstructionAnalyzer.cpp#L674 — what else is this, other than a jump?

**Comment by hlide** [comment]
ahh yes
```c++
    case I_SLTI: {
        AllegrexExpressionFactoryCallback _1(factory, program->createBasicBlock(), instruction);
        AllegrexExpressionFactoryCallback _0(factory, program->createBasicBlock(), instruction);
        _[
            jump((signed_(gpr(1)) < signed_(imm(2))),
                 _1[gpr(0) ^= constant(1), jump(directSuccessor())].basicBlock(),
                 _0[gpr(0) ^= constant(0), jump(directSuccessor())].basicBlock())
        ];
        break;
    }
```

**Comment by hlide** [comment]
nasty one... any suggestion ?

**Comment by yegord** [comment]
By the way, there is a simpler implementation of SLTI: 
```c++
        _[
            gpr(0) ^= signed_(gpr(1)) < signed_(imm(2))
        ];
```
IMHO, you overuse jumps heavily.

**Comment by hlide** [comment]
this one is setting 0 or 1 according to operator <?

**Comment by yegord** [comment]
Yes. Comparison operators return 1-bit value, where 1 stands for true, 0 for false.

**Comment by hlide** [comment]
ok so it is a better solution as it is exactly what this instruction is doing. But I can other issue...

**Comment by yegord** [comment]
But in the general case, when a delayslot instruction has complex control flow, you have to invent something. I leave it to you as an exercise for the night. :-)

**Comment by hlide** [comment]
*MOVN* and *MOVZ* for instance

**Comment by hlide** [comment]
So, the issue seems to be around:
```c++
auto then = delayslot(AllegrexExpressionFactoryCallback(factory, program->createBasicBlock(), instruction));
then[jump(imm(2))];
_[jump(gpr(0) == gpr(1), then.basicBlock(),  directSuccessor())]; // BEQ
```
for some weird reason, `jump(imm(2))` is not appended to *then* basic block but pushed to the end of *BEQ* basic block.

**Comment by hlide** [comment]
My intent was to have:

```
{0}: if (C) then goto {1} else goto {2}

{1}: D; goto target

{2}: ....
```

but I get:
```
{0}: if (C) then goto {1} else goto {2}; goto target

{1}: ???

{2}: ....
```

**Comment by hlide** [comment]
so I must move this jump inside {3}. That is, the most nested jump should get this delayslot jump in its then and else basic blocks.

**Comment by yegord** [comment]
You should write a function F that takes an instruction, the basic block to which the first statement implementing this instruction must be added, and the basic block that must get control after the instruction is executed.

To generate code for a usual instruction, you call
```c++
F(instr, program.getBasicBlockForInstruction(instr), program.createBasicBlock(instr->endAddr()) /*(!)*/)
```

To generate code for a delay-slot instruction, you call
```c++
auto successor = program.createBasicBlock();
F(delayslotInstr, program.getBasicBlockForInstruction(jump), successor)
```
. The code of `jump` is then added to `successor`. And at the end of successor you jump to `program.createBasicBlock(jump->endAddr())`.

Of course, you need to optimize this solution a bit, in order to not to create a basic block per instruction due to (!), but this should be easy.

**Comment by hlide** [comment]
To sum up, for this case:
```
@0: beqz $v0, @target
@1: movn $v1, $a0, $a1
@2: ...
```
I currently have:
```
[
	jump
	(
		v0 == 0,
		[
			jump
			(
				v1 != a1,
				[
					v1 ^= a0,
					goto @2
				],
				[
					goto @2
				]
			),
			goto @target
		],
		[
			goto @1
		]
	)
]
```
but I must have:
```
[
	jump
	(
		v0 == 0,
		[
			jump
			(
				v1 != a1,
				[
					v1 ^= a0,
					goto @target
				],
				[
					goto @target
				]
			)
		],
		[
			goto @1
		]
	)
]
```

where `goto @2` must be replaced with `goto @target` and the original `goto @target` removed.

**Comment by hlide** [comment]
*sigh* A more complex one:
```
@0: bltzal $v0, @target
@1: movn $v1, $a0, $a1
@2: ...
```
currently:
```
[
    jump
    (
        v0 < 0,
        [
            jump
            (
                v1 != a1,
                [
                    v1 ^= a0,
                    goto @2
                ],
                [
                    goto @2
                ]
            ),
            call @target,
            goto @2
        ],
        [
            goto @1
        ]
    )
]
```
correct one:
```
[
    jump
    (
        v0 < 0,
        [
            jump
            (
                v1 != a1,
                [
                    v1 ^= a0,
                    [
                            call @target,
                            goto @2
                    ]
                ],
                [
                    [
                            call @target,
                            goto @2
                    ]
                ]
            )
        ],
        [
            goto @1
        ]
    )
]
```

**Comment by nihilus** [comment]
@hlide: for the register comparsion you in JR, could you please implement getOperandRegister to conform with ARM and MIPS and convience. :-) I guess that would help other people implement their platforms.

**Comment by nihilus** [comment]
@yegord: thx for the suggestion on slti(u).

**Comment by hlide** [comment]
@yegord 

I changed:
```c++
void createStatements(AllegrexExpressionFactoryCallback & _, const AllegrexInstruction *instruction, core::ir::Program *program);
```
into:
```c++
void createStatements(AllegrexExpressionFactoryCallback & _, const AllegrexInstruction *instruction, core::ir::Program *program, core::ir::BasicBlock * successor);
```
I took into consideration the optimization you talked about by having `successor == nullptr` for an usual instruction so that a `program->createBasicBlock(instruction->endAddr())` is done only once per call to `createStatements` and only if the usual instruction needs it.

So, if we look for `MOVN`, we have:
```c++
    case I_MOVN: {
        auto succ = successor ? successor : directSuccessor();
        auto then = AllegrexExpressionFactoryCallback(factory_, program->createBasicBlock(), instruction)[
            gpr(0) ^= gpr(1), jump(succ)
        ];
        auto none = AllegrexExpressionFactoryCallback(factory_, program->createBasicBlock(), instruction)[
            jump(succ)
        ];
        _[
            jump(gpr(2) == constant(0),
                 then.basicBlock(),
                 none.basicBlock())
        ];
        break;
    }
```
since `MOVN` is a leaf instruction (no delay slot), it is safe to issue here a `jump(succ)`. If `MOVN` is  in the delay slot of a branch instruction, so it means `succ`is a basic block which can contain one amongst:

* `[jump(<target>)]` from *J, JR, Bxx, BxxL*

* `[call(<target>), jump(<direct successor but one>)]` from *JAL, JALR, BxxAL, BxxALL*

* `[jump(return_address())]` from *JR $RA*
 
Now let us consider what we need to make in a *branch delay slot*:

1) fetch the delay slot instruction and call `createStatements` for it with the created basic block `succ` according to the type of the branch instruction.

2) check if the delay slot basic block ends with a terminator (example: `MOVN` does):

2.1) it does: no need to issue `jump(succ)` after calling `createStatements`. The delay slot already did it by propagating `succ`in its *then* and/or *else* basic blocks .

2.2) it does not: let us append `jump(succ)` in the delay slot basic block to make it this delay slot a terminator.

What do you think about?

**Comment by hlide** [comment]
@nihilus: ALLEGREX decomposer - unlike MIPS capstone decomposer - has no ambiguity between operands types and the instruction ID: you know exactly if such operand is a *gpr*, a *fpr*, an *imm*, etc. by just having the right ID. It makes more readable and dynamic checking is useless in this case. As far as I know there are big differences between x86[-64], ARM and MIPS implementations even they have in common the use of `operand(n)`, so I am not sure I will gain to do so.

**Comment by nihilus** [comment]
@hlide: Ah so the getOperandRegister would be redudant you mean? Yes ofc it will be, but for the novice implementer easier to understand IMHO.

**Comment by yegord** [comment]
I think that your new `createStatements` should just return a pointer to the basic block to which the instruction with the delay slot should add its statements. Simple instructions just return the basic block to which they added statements, complex instructions like MOVN return `succ`.

**Comment by yegord** [comment]
Okay, perhaps I should just write code. :-)

**Comment by nihilus** [comment]
@yegord you have the best understanding of internals atm, so be our Guest. :-)

**Comment by hlide** [comment]
Ok I think I grab your proposal:

1) the branch instruction will get the basic block returned by the delay slot instruction to append the necessary statements:

* `[jump(<target>)]` from *J, JR, Bxx, BxxL*

* `[call(<target>), jump(<direct successor but one>)]` from *JAL, JALR, BxxAL, BxxALL*

* `[jump(return_address())]` from *JR $RA*

2) a simple instruction with no control flow just returned its won basic block.

3) a complex instruction with control flow will create a new basic block and returns it as `succ`. Before returning it, the complex instruction will a `jump(succ)` inside *ThenStatement* and/or *ElseStatement*.

Yeah, it sounds easier and simpler to implement. :-)

**Comment by hlide** [comment]
3) no need to create `succ` it is done by 1) indeed

**Comment by yegord** [comment]
Something like this: https://github.com/yegord/snowman/commit/1732a174032a5705fc7609875ac8d7c9887159a0

**Comment by hlide** [comment]
ok you're puting a side effect on `_` by switching basic block. While I understand how it happens with 'J', I have some hard job with conditional branch instructions. I cannot just do:
```c++
                        _[
                            jump(gpr(0) == gpr(1),
                                 delayslot.basicBlock(),
                                 directSuccessor())
                        ];
```
after calling the delayslot function which switch basic block of `_`.

**Comment by hlide** [comment]
*sigh* it fails on it:
```c++
    /*
     * Check statements are sorted by their instructions' addresses.
     * ir::Program::createBasicBlock(ByteAddr) relies on this while splitting basic blocks.
     */
```

**Comment by yegord** [comment]
On Sun, Aug 09, 2015 at 08:51:10AM -0700, hlide wrote:
> I cannot just do:
> ```c++
>                         _[
>                             jump(gpr(0) == gpr(1),
>                                  delayslot.basicBlock(),
>                                  directSuccessor())
>                         ];
> ```

Why do you want to jump to delayslot.basicBlock()?

-- 
Yegor Derevenets

**Comment by hlide** [comment]
Well, the issue is more I do not how to rewrite it. And because the fact `_` can be altered it make less understandable what happens internally and so I decided not to allow it but still keeping the same idea as yours (I explicitly give the callback I want to assign the basic block returned by the createStatements on the delay slot to keep a full control on them)

**Comment by hlide** [comment]
Good news:

![image](https://cloud.githubusercontent.com/assets/56785/9156217/4b7b1ce8-3ed2-11e5-8511-bf1abc2515e7.png)

Functions appear less void but I cannot tell if they are all correct.

**Comment by nihilus** [comment]
@hlide how does the Allegrex way differ from the MipsInstructionAnalyzer.cpp way?

**Comment by hlide** [comment]
1) cleaner instruction decomposer unlike capstone: instruction ID are really unique par opcode and no ambiguity with operand types. Aliases may or not used - right now, not used.
2) full disassembler supporting instructions aliases for display purpose.
3) each instruction id has its own case.
4) each instruction id emits IR expression as it is supposed to run for Allegrex (unimplemented ones are set as inline statement - while I would prefer *intrinsic functions* instead).

**Comment by nihilus** [comment]
@hlide I was thinking about delay slots to be specific.

**Comment by hlide** [comment]
Allegrex disassembler is not so different from gnu libutils so you may indeed get something similar to Allegrex disassembler for MIPS.

**Comment by hlide** [comment]
And you could indeed reuse a lot of things in AllegrexAnalyzer for MIPS with a similar decomposer (to code as I needed to do for Allegrex - unless libbfd is able to so).

**Comment by nihilus** [comment]
Yeah, I was sleeping / hangover too much this weekend :-( so boo tries with libbfd actually. :-1: 
Now I am trying to figure out what's wrong with the SWR/SWL/LWL/LWR stuff.. Seems like 'memval' is incorrect.

**Comment by hlide** [comment]
well in 99% of cases, you have a pair LWL/LWR to make a pseudo ULW (unaligned load word), so having a complex IR expression for LWL and then for LWR  is madness just to express an unaligned load. That is why I would like intrinsic functions for them:

```
LWL rd, offset+3(rs) --> rd = __allegrex_lwl(rd, rs, offset+3);
LWR rd, offset+0(rs) --> rd = __allegrex_lwr(rd, rs, offset+0);

LWL rd, offset+3(rs) --\
                       +-->  rd = __allegrex_lwr(__allegrex_lwl(rd, rs, offset+3), rs, offset+0);
LWR rd, offset+0(rs) --/               |
                                       `--> rd = __allegrex_ulw(rs, offset);
```

**Comment by nihilus** [comment]
Yes, this is would be nice to transform it into ulw / usw.

**Comment by hlide** [comment]
on the other hand, I wonder if my disassemble/decomposer should be able to handle a sequence of two instructions with one instruction ID. Snowman is able to handle instructions with variable size. So I could handle LA and ULW, and USW with  an instruction ID and a 8-byte size instead of the usual 4-byte size.

**Comment by nihilus** [comment]
Is there any easy way to log what 'auto memval = *(ea & constant(-4));' actually is?

**Comment by hlide** [comment]
`memval.toString()` ?

**Comment by yegord** [comment]
Concerning commenting out well-formedness checks — fix your code. :)
Concerning LWL/LWR pairs, I would make a separate pass transforming them into usual loads. (Similar to how implicit zero-extend on x86 is handled.)
Concerning memval, you can make an IR statement with it, using the usual [] stuff, and then print the statement, using operator<< or Statement::toString().

**Comment by hlide** [comment]
oh you mean to output on log window, I added some stuff to do so...

```diff
From 6d67bbf9164ac422dba55a009e0865217ae6a53c Mon Sep 17 00:00:00 2001
From: hlide <hlide@free.fr>
Date: Mon, 10 Aug 2015 01:03:04 +0200
Subject: [PATCH] Added log facility

---
 src/nc/common/LogToken.h  | 2 ++
 src/nc/gui/MainWindow.cpp | 2 +-
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/nc/common/LogToken.h b/src/nc/common/LogToken.h
index 01a705b..2135e6c 100644
--- a/src/nc/common/LogToken.h
+++ b/src/nc/common/LogToken.h
@@ -40,6 +40,8 @@ class LogToken {
     std::shared_ptr<Logger> logger_;
 
 public:
+    static LogToken* instance(LogToken* token = nullptr) { static LogToken* token___ = nullptr; if (token && !token___) token___ = token; return token___; }
+
     /**
      * Default constructor.
      *
diff --git a/src/nc/gui/MainWindow.cpp b/src/nc/gui/MainWindow.cpp
index 082c1f6..0450595 100644
--- a/src/nc/gui/MainWindow.cpp
+++ b/src/nc/gui/MainWindow.cpp
@@ -86,7 +86,7 @@ MainWindow::MainWindow(Branding branding, QWidget *parent):
     connect(logger.get(), SIGNAL(onMessage(const QString &)), progressDialog_, SLOT(setLabelText(const QString &)));
     connect(logger.get(), SIGNAL(onMessage(const QString &)), this, SLOT(setStatusText(const QString &)));
 
-    logToken_ = LogToken(logger);
+    LogToken::instance(&(logToken_ = LogToken(logger)));
 
     settings_ = new QSettings(branding_.organizationName(), branding_.applicationName(), this);
     loadSettings();
-- 
1.9.4.msysgit.2
```

**Comment by nihilus** [comment]
ah, thx

**Comment by yegord** [comment]
You could just use qDebug() — it's output is redirected to log window.

**Comment by hlide** [comment]
@yegord I'm not sure I can fix the issue with sorted addresses because you need to put a delay slot instruction *(address + 4) **BEFORE** the real branch instruction *(address)*.

**Comment by nihilus** [comment]
@yegord I've also started to get 'Assertion failed: (statement == basicBlock->statements().back()), function generate, file /Users/nietzsche/Downloads/snowman/src/nc/core/irgen/IRGenerator.cpp, line 110.'

**Comment by hlide** [comment]
@nihilus the issue you have is almost the same I had with Allegrex.

**Comment by yegord** [comment]
@hlide You need to mark the statements of the delay slot instruction that you put before the statements of the delay slot owner as belonging to the delay slot owner. In my proposed patch I do that.

@nihilus This is good. Fix your code.

**Comment by hlide** [comment]
You proposed patch did not work so I reverted while keeping some ideas of yours. But I will try the thing about delay slot owner.

**Comment by nihilus** [comment]
@yegord and it means that my blocks in jumps are bogus???

**Comment by hlide** [comment]
@nihilus when sober (:-)), just read above all the posts regarding issue with branch instructions - they are the same for MIPS.

**Comment by hlide** [comment]
Good news. With all the checking activated, it passes now.

**Comment by yegord** [comment]
Just a small comment on why statements in a basic block should be sorted by instruction's addresses.
IRGenerator sometimes wants to split an existing basic block, when there is a jump into the middle of it.
So, if you have a basic block with instructions with addresses 1, 2, 3, 2, 1 and a jump to address 2, it is not very clear, how to split the basic block, and what the boundaries of the new two basic blocks should be.
I think, in case of MIPS, you can only have something like 1, 2, 4, 3, and the current implementation of splitting will de-facto work, but I would prefer to stay on the side where things are simple.

**Comment by uxmal** [comment]
You've probably studied this already, but here is a paper [http://www.cs.tufts.edu/~nr/pubs/xtoplas-acm.pdf] details how to cope with delay slots. I believe this is what the boomerang folks based their implementation on.

**Comment by yegord** [comment]
Hmm. In the paper they even try to give semantics to jumps in delay slots, something that seems to be not needed for MIPS: http://electronics.stackexchange.com/questions/28444/mips-pic32-branch-vs-branch-likely

**Comment by hlide** [comment]
@yegord that is fine, I know how to handle correctly.

As for Allegrex, you may find some games using a branch instruction inside - surprise ! - a branch instruction. I know what happens but that's specific to Allegrex which has a 7-stage pipeline. *Bxx[L]* has two bubbles (three cycles of latency) and *Jxx* has one bubble (two cycles of latency). If both are taken-able, the second is taken for all combinations  except for one : 

```python
def interpret_delay_slot(pc, is_cond):
    insn = fetch(pc)    
    if insn.is_cond_branch() is True:
        if insn.test_cond_branch() is True:
            return insn
        else:
            return None
    elif insn.is_uncond_branch() is true:
        if is_cond is false:
            return insn 
        else:
            return None
    else:
        ...

def interpret(pc, delay_slot):
    insn = fetch(pc)
    if insn.is_cond_branch() is True:
        if insn.test_cond_branch() is True:       
            if delay_slot is False:
                delay_slot_insn = interpret_delay_slot(pc + 4, True)
                if delay_slot_insn is not None:
                    return delay_slot_insn.target_branch()
            return insn.target_branch()
        else:
            return pc + 4
    elif insn.is_uncond_branch() is true:
        if delay_slot is False:
            delay_slot_insn = interpret_delay_slot(pc + 4, False)
            if delay_slot_insn is not None:
                return delay_slot_insn.target_branch()
        return insn.target_branch()
    else:
        ...
```

1) All conditional branch instructions included COP1 and COP2 have one delay slot + 2 bubbles when branch is taken.

1.1) When not taken, the next instruction is executed as a normal instruction, not as delay slot instruction.

1.2) When taken, if its delay slot instruction is also a conditional branch instruction which is taken, only the second instruction jumps to its target.

1.3) When taken, if its delay slot instruction is a unconditional branch instruction, only the first instruction jumps to its target.

2) All unconditional branch instructions have one delay slot + 1 bubble when branch is taken.

2.1) When not taken, the next instruction is executed as a normal instruction, not as delay slot instruction.

2.2) When taken, if its delay slot instruction is a branch instruction which is taken, only the second instruction jumps to its target.

**Comment by nihilus** [comment]
@uxmal: No we did not based boomerang upon that paper... Every machine is represented in a language called SLED invented by Norman Ramsey and the NJ Machine Code Toolkit as found here: https://www.cs.tufts.edu/~nr/toolkit/

from there everything will get an immediate representation which is basically a minimalistic Turning-machine. Etcetc.

**Comment by nihilus** [comment]
@hlide turned out I got a cut 'n paste error in swl for BE. :-) Thx @yegord  for adding those extra checks.

**Comment by nihilus** [comment]
@yegord 'Concerning memval, you can make an IR statement with it, using the usual [] stuff, and then print the statement, using operator<< or Statement::toString().'

Could you please give me a hint. I included QDebug etc... But how to convert memval to a statement :-/?

**Comment by hlide** [comment]
Interesting stuff. Is there a git-mirrored Boomerang repository? and for this toolkit? they may give some ideas for a multi-platform assembler/disassembler with a focus upon composer/decomposer.

**Comment by hlide** [comment]
#### OFF TOPIC MODE:on

I was starting a run-time assembler with kinda of composer in pure c++ template [here](https://github.com/hlide/jitasm) but x86 is a horrid beast to tame :(.

#### OFF TOPIC: off

**Comment by uxmal** [comment]
There are _many_ Boomerang repositories on github, here's one: https://github.com/nemerle/boomerang

**Comment by nihilus** [comment]
@uxmul yes but the official site is at http://boomerang.SF.net however I am looking for a new admin.

**Comment by uxmal** [comment]
Is any active development being done on boomerang? I see a lot of punters cloning the project but it seems to me that they're mostly doing code polishing, but no actual development.

**Comment by nihilus** [comment]
@uxmal: I asked nermerle to take over but he didnt respond. No I dont do anything activly on it.

It is very accurate but unstable and has got no smart pointers etc.

IMO using Capstone with Snowman is much easier to get things working since the NJMC ML is outdated and I cannot get it running natively on OS X. Ramsey is too occupied to care about it. Etc etc.

**Comment by nihilus** [comment]
@hlide see @uxmal :-) Yes, boomerang have some nice features like function recognition based on C-header files.


There is also a branch for a plugin-API, which would be nice to have for snowman as well.

However it does implement it own object parsers like Snowman instead of reusing code.

There is a static linux binary for NJML / NJMC and some MIPS-stubs there which wrote ages ago.

**Comment by yegord** [comment]
On Mon, Aug 10, 2015 at 02:56:13PM -0700, Markus Gothe wrote:
> @yegord 'Concerning memval, you can make an IR statement with it,
> using the usual [] stuff, and then print the statement, using
> operator<< or Statement::toString().'
> 
> Could you please give me a hint. I included QDebug etc... But how to
> convert memval to a statement :-/? 

Something like:

_[*zero ^= memval];
qDebug() << _.basicBlock()->statements()->back()->asAssignment()
	->right()->toString();

Or, without adding to the basic block:

qDebug() << factory.createStatement(*zero ^= memval)
	->asAssignment()->right()->toString();

-- 
Yegor Derevenets
